### ISSUE TITLE
Feedback on the editor inline hints implementation

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#127703]
This is some generic feedback on the editor inline hints implementation, not this TS-specific support, but I wasn’t sure where else to share it. I found the experience of splitting up a long function call with parameter name hints extremely disorienting, because I wanted to put my cursor before the hint to move it down a line with the argument. But the cursor position only exists _after_ the hint, so the impression you get as you make a refactor like this is that you’re actually inserting a newline between the label and the argument it’s labeling. And indeed, there is a moment of delay before the label disappears and reappears on the new line:

https://user-images.githubusercontent.com/3277153/122442211-f12e7e00-cf63-11eb-80c8-4863b1a7c710.mp4

I obviously understand what’s happening and how I need to adjust for it—I’m exaggerating my struggle in this screen cap to demonstrate the nature of why it feels unintuitive—but it really is kind of jarring. It seems to me that you’d either want to duplicate the cursor position on either side of these labels, so it appears as if there are two cursor positions, even though they’re actually the same position in the source text, _or_ you’d want an API that allows these labels to be “sticky” on either the start or end, which would control which side of the label you can place the cursor. Prefix labels like these would be sticky on their end side, such that you can’t separate the label from the argument that follows it. This would result not only in a more intuitive editing experience, but also fewer reflows of label position mid-edit like you see in every newline insertion here.

_Originally posted by @andrewbranch in https://github.com/microsoft/vscode/issues/113412#issuecomment-863413819_

**Comment by jrieken** [comment]
> . It seems to me that you’d either want to duplicate the cursor position on either side of these labels, so it appears as if there are two cursor positions, even though they’re actually the same position in the source text

💯 that's basically why we haven't moved forward with this. Today, inlay hints are implemented with ::before/after pseudo elements and that has all kind of funny and annoying issues. @hediet and @alexdima work on true support for inline text and that should resolve these kind of problems

**Comment by Kingwl** [comment]
Very interested in this, please let me know how I can help If I may.

**Comment by hediet** [comment]
We are already working on something, but thanks @Kingwl!
It should land in VS Code Insiders in the next couple of days.

**Comment by Kingwl** [comment]
Cool! Thanks!

**Comment by Kingwl** [comment]
I guess it's https://github.com/microsoft/vscode/commits/hediet/injected-text-position-normalization this branch?  Very looking forward on it.

**Comment by Kingwl** [comment]
I guess https://github.com/microsoft/vscode/pull/128140 has fixed this?

**Comment by hediet** [comment]
not yet, the new API needs to be used. I'm on it.

**Comment by yume-chan** [comment]
![image](https://user-images.githubusercontent.com/1330321/126253475-78ef0615-91d3-4d68-b931-f3425601d2c1.png)

TypeScript inlay hints has switched to the new "true element" mode!

> It seems to me that you’d either want to duplicate the cursor position on either side of these labels, so it appears as if there are two cursor positions, even though they’re actually the same position in the source text,

Now there ARE cursor stops on both sides of inlay hints, but when moving cursor with arrow keys, it only stops at one of them depending on which direction the cursor is moving.

https://user-images.githubusercontent.com/1330321/126256159-3f68b091-2cf2-4ed0-92cf-6727288ff7f3.mp4

I think it should stop on both, as if these inlay hints are real text in editors.

**Comment by hediet** [comment]
> I think it should stop on both, as if these inlay hints are real text in editors.

IntelliJ does this.
There are some problems though:
* When using multiple cursors, moving left/right would not move the text position of all cursors if some of them are before/after inlay hints. This might be unexpected.
* Sometimes inlay hints need some time to be computed and appear only after some delay. This could happen while the users navigates in the line and could be annoying.

The current behavior ensures that cursor navigation without and with inlay hints is the same.

However, I'm open to change this.

**Comment by jrieken** [comment]
> However, I'm open to change this.

Yeah, I think that should be tried. I believe there is desire to have these virtual stops before and after the inlay text, at least as an option

**Comment by hediet** [comment]
What about line wrapping points?
When a line is wrapped, there are also two cursor stops that point to the same model position.
Currently, we skip one of those (which is important when using multiple cursors).

**Comment by seansfkelley** [comment]
The multiple stop points is nice, but I find myself doing this more often than I'd like:

https://user-images.githubusercontent.com/2101409/135725276-949c2c5b-aaa2-4fd0-aa3e-1adab412f2ea.mov

If it's not clear: I'm arrowing over from the right side of a function return type inlay hint in an attempt to get to the left side of the inlay so I can add an explicit type annotation (`: boolean`, in this case). But when you come from the right side, it stops only at the stop point to the right of the inlay. I would either have to stop there and write `: boolean` (which feels wrong, because there's a thing in between the cursor and the `)` which I am trying to append my changes to) or go past the desired insertion point and approach the hint from the left side (which I don't think to do, since I'm coming from the right side).

**Comment by OldStarchy** [comment]
I think using a single stop, and having it highlight the hint like @seansfkelley's mockup (regardless of cursor type) makes it more intuitive.

If you do move your cursor to that position, intuition tells me that you're likely about to do something that is going to invalidate the hint anyway (again see video above). Typing then would remove the inlay hint and it wouldn't return until you stop typing (#133730).

Doing this over having two "virtual" stops naturally avoids the issues mentioned above.

**Comment by 0b1100yh** [comment]
> Doing this over having two "virtual" stops naturally avoids the issues mentioned above.

currently, actual stop is the left "virtual" stop, it caused #133385，we can resolve it by choosing right "virtual" stop as actual stop, and this makes the cursor movement more visually reasonable.

**Comment by 0b1100yh** [comment]
Having just downloaded the IntelliJ and tried it out, the cursor stops to the right of the inlay hint instead of the left, which is exactly what I suggested above. Please fix it (I hope to catch up with 1.68 recovery if possible). This problem makes my use experience very bad.

**Comment by ifyuhj** [comment]
> currently, actual stop is the left "virtual" stop, it caused #133385，we can resolve this by choosing the right "virtual" stop as actual stop, and it makes the cursor movement more visually reasonable.

Hello, Is it possible to display the cursor at correct position for vim users? 
If non vim users think the current scheme is better，we can add an option like this: 
```jsonc
inlayhints.cursorPostion: "left" // or right
```

**Comment by KurtGokhan** [comment]
Is there a consensus on this? If not, I would like to provide my thoughts.

Usually, inlay hints are either a prefix, or suffix. For instance, when it is a suffix, you usually want to put caret stop on the start. I think having an option like this would probably please everyone.

- `auto` (or `directional`, The current behavior. The caret stop is skipped depending on the direction of caret navigation)
- `both` (There is a caret stop on both sides and they are not skipped)
- `start` (Caret stop is always on start. End is always skipped)
- `end`(Opposite of `start`)

**Comment by ant1fact** [comment]
Hello, may I ask if there has been any further progress on this? Thanks
